src/uct/ib/base/ib_iface.c

```diff
@@ -245,59 +245,65 @@ size_t uct_ib_address_size(uct_ib_iface_t *iface)
-                         const union ibv_gid *gid, uint16_t lid,
+size_t uct_ib_iface_address_size(uct_ib_iface_t *iface)
+{
+    return uct_ib_addr_size(&iface->gid, iface->is_global_addr, uct_ib_iface_is_roce(iface));
+}
+
+void uct_ib_address_pack(const union ibv_gid *gid, uint16_t lid,
                          uct_ib_address_t *ib_addr)
 {
     void *ptr = ib_addr + 1;
 
-    if (uct_ib_iface_is_roce(iface)) {
+    if (ucs_test_all_flags(ib_addr->flags, UCT_IB_ADDRESS_FLAG_LINK_LAYER_ETH |
+                                           UCT_IB_ADDRESS_FLAG_GID)) {
         /* RoCE, in this case we don't use the lid and set the GID flag */
-        ib_addr->flags = UCT_IB_ADDRESS_FLAG_LINK_LAYER_ETH |
-                         UCT_IB_ADDRESS_FLAG_GID;
         /* uint8_t raw[16]; */
         memcpy(ptr, gid->raw, sizeof(gid->raw) * sizeof(uint8_t));
-    } else {
-        /* IB, LID */
-        ib_addr->flags = UCT_IB_ADDRESS_FLAG_LINK_LAYER_IB |
-                         UCT_IB_ADDRESS_FLAG_LID;
-        *(uint16_t*) ptr = lid;
-        ptr += sizeof(uint16_t);
+        return;
+    }
 
-        if ((gid->global.subnet_prefix != UCT_IB_LINK_LOCAL_PREFIX) ||
-            iface->is_global_addr) {
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_IF_ID;
-            *(uint64_t*) ptr = gid->global.interface_id;
-            ptr += sizeof(uint64_t);
-
-            if (((gid->global.subnet_prefix & UCT_IB_SITE_LOCAL_MASK) ==
-                                              UCT_IB_SITE_LOCAL_PREFIX) &&
-                !iface->is_global_addr) {
-                /* Site-local */
-                ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET16;
-                *(uint16_t*) ptr = gid->global.subnet_prefix >> 48;
-            } else {
-                /* Global */
-                ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET64;
-                *(uint64_t*) ptr = gid->global.subnet_prefix;
-            }
-        }
+    /* IB, LID */
+    ucs_assert(ucs_test_all_flags(ib_addr->flags,
+                                  UCT_IB_ADDRESS_FLAG_LINK_LAYER_IB |
+                                  UCT_IB_ADDRESS_FLAG_LID));
+    *(uint16_t*) ptr = lid;
+    ptr += sizeof(uint16_t);
+
+    if (!(ib_addr->flags & UCT_IB_ADDRESS_FLAG_IF_ID)) {
+        return;
+    }
+    *(uint64_t*) ptr = gid->global.interface_id;
+    ptr += sizeof(uint64_t);
+
+    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_SUBNET16) {
+        /* Site-local */
+        *(uint16_t*) ptr = gid->global.subnet_prefix >> 48;
+        return;
     }
+
+    ucs_assert(ib_addr->flags & UCT_IB_ADDRESS_FLAG_SUBNET64);
+    /* Global */
+    *(uint64_t*) ptr = gid->global.subnet_prefix;
 }
 
 void uct_ib_address_unpack(const uct_ib_address_t *ib_addr, uint16_t *lid,
                            union ibv_gid *gid)
 {
     const void *ptr = ib_addr + 1;
 
-    gid->global.subnet_prefix = UCT_IB_LINK_LOCAL_PREFIX; /* Default prefix */
-    gid->global.interface_id  = 0;
+    memset(gid, 0, sizeof(*gid));
```

Comment:
why need to set it to 0 here?

Reply:
for initialization before filling it, like *lid is set to 0 the line after.

Reply:
but gid is overridden anywan in lines 299, and 304-305